### ISSUE TITLE
[Addon: A11y] - Feature - Highlight the violation on the rendered component

### ISSUE THREAD

**Issue by CodeByAlex** [storybooks/storybook#6001]
**Is your feature request related to a problem? Please describe.**
There isn't a way for a non-technical user to determine where the accessibility issue is located on the component. I think it would be useful for there to be a button/checkbox/link that, when clicked, would highlight where the specific issue is located on the component.

**Describe the solution you'd like**
I would like there to be a way for a user to go to the a11y tab, look at a violation and be able to pinpoint where that issue is on the component using some form of highlighting.
 ![image](https://user-images.githubusercontent.com/29763590/54087527-10bcb200-432a-11e9-8d5c-5bc6dc4bc6c6.png)

**Implementation idea**
Run querySelector from the root of the iframe using the item -> nodes -> target (which provides a path to the element) once the elment is found, it can be highlighted. This code will be triggered on some type of action depending upon design

**Are you able to assist bring the feature to reality?**
Yes

**Additional context**
@domyen would you want to collaborate on this so that we can maintain a consistent design

**Comment by shilman** [comment]
That's a cool idea! I wonder whether we can do this without modifying the component's DOM? @ndelangen maybe a new addon API?

**Comment by CodeByAlex** [comment]
@shilman I like the idea of keeping the changes out of the component’s DOM if possible. Any idea  how that could be done whether it’s a new api or not?

**Comment by shilman** [comment]
AFAIK it's not possible in the current setup. My suggestion would be prototyping the feature by modifying the DOM directly and then potentially rebuild it with a new API if we can figure out a good way to do that.

**Comment by CodeByAlex** [comment]
Nice @Armanio! So are you saying that you are already working on adding this feature or are you saying that you would like this feature to be added?

**Comment by Armanio** [comment]
Yep, i already started work on this feature. But dont worry, if you want you can implement it instead me.

**Comment by CodeByAlex** [comment]
I started working on this last night so I'm not too far along yet. How much progress have you made?

In any case, at some point, I think we should touch base with @ndelangen to see if he has any ideas for the design. Were you planning on making the path a link, adding a button, or something else to trigger the highlighting

**Comment by domyen** [comment]
I can take a look at the design 👍 Socialize the prototype when ready, also
lmk the outcome of the api discussion :)

**Comment by Armanio** [comment]
@CodeByAlex i write prototype but code is too bad. 😖
I am ashamed to show it.

I recorded mini screencast.
![2019-03-11 23 46 16](https://user-images.githubusercontent.com/3195714/54156959-5d9ea680-4458-11e9-9e58-d520708e8b1b.gif)

**Comment by CodeByAlex** [comment]
@Armanio Nice work! I like the idea of having transparent styles added to where the accessibility rule is checking. The user can see the component through the highlight and, since it is temporary, it means that the component is not altered with a border. However, I am wondering if it would be frustrating for the user to have to keep clicking in order to see the location of the rule. Maybe we can have the transparent glow last longer or have it on a toggle that can be turned on or off. What do you think?

**Comment by Armanio** [comment]
Now i add 'pulse' class to target element for 2 sec on click. We can do it on hover. On 'mouseenter' - add class, on 'mouseleave' - remove class. So user can turn on/of glowing manually.  🤔

**Comment by domyen** [comment]
@Armanio This is an awesome animation for calling attention to an element! 

Agree with @CodeByAlex, a persistent highlight would be ideal. 

Strawman UX:
- Add highlight to elements that are in violation `outline: 1px dotted red !important` [1]
- All highlights are visible by default when the accessibility tab is activated 
- Add checkbox in the accessibility addons panel subheader to toggle between showing the highlight and hiding the highlight (by default it is checked)
- The animation fires only when the highlight is toggled


[1] This calls attention to the violating element without changing the rendered layout (similar to [Pesticide Chrome extension](https://chrome.google.com/webstore/detail/pesticide-for-chrome/bblbgcheenepgnnajgfpiicnbbdmmooh?hl=en-US))

**Comment by Armanio** [comment]
I understand! You want highlight violation element right away! Without user actions! Of course! Excuse me, today i cant think straight. 🙃

**Comment by shilman** [comment]
This is so freaking cool. Great job @Armanio @CodeByAlex -- this is gonna be an awesome feature for a11y.

**Comment by CodeByAlex** [comment]
@domyen thanks for the strawman design! That will definitely help improve the experience and increase accessibility in the process by having a persistent visual rather than a flash.

What do you all think about providing the same functionality for passing (green) and incomplete (yellow). When you move from tab to tab, it should only be able to highlight the rules within that tab. This would add a consistent feel across all tabs.

@armanio I feel like this change would require minimal additional work as you could reuse the code that pinpoints the element in the dom and  modify the color based on type. Not sure how you’re breaking up the code though @armanio. What do you think?

**Comment by Armanio** [comment]
Green border for passed elements, red for violations, orange for incomplete? Sounds good!

I get iframe by id, after in 'contentDocument' get by query from rule. Its all. :wink:

Without keyframe animation, additional class and onclick event its look a like easy feature. 
@CodeByAlex Could you implement that?

**Comment by CodeByAlex** [comment]
Yeah I can take this on. Thanks for getting a POC going @Armanio

**Comment by CodeByAlex** [comment]
> * Add checkbox in the accessibility addons panel subheader to toggle between showing the highlight and hiding the highlight (by default it is checked)

@domyen How would the user know which rule corresponds to which highlighted option if the only options that they have are to show all highlighting or show no highlighting

**Comment by domyen** [comment]
@CodeByAlex good question! Armanio's flash animation would be the right tool for this. 

When the highlights are shown, you can click on a list item to trigger the animation as shown in the GIF.

The other thought I've been considering is to reveal the a11y message on hover. 
Something like this:
![image](https://user-images.githubusercontent.com/263385/54223391-c3794580-44cd-11e9-859a-dd0dc2e7acf0.png)

This would be a better DX because you wouldn't have to cross reference between the panel and the real UI. The tradeoff is that it pollutes the component DOM in unpredictable ways. What do you think?

**Comment by CodeByAlex** [comment]
@domyen I like the idea of having the description popup as it gives insight into the issue while the user is looking at the component. This way they won't have to look back and forth between the addon view and the component. However, hover states don't work on mobile so we might have to figure out how to make a good mobile experience.  A click is not really an option because then you would be interfering with the UI. Ex the button is the issue - if you click the button, the component may react which would not be our intent

**Comment by CodeByAlex** [comment]
@domyen so here's what I'm thinking. Let's take an iterative approach to get this feature released.

First iteration: Add a toggle button next to each of the rules that allow a user to add or remove the highlight border for the rule. The color of the border for each rule should change based on type (success, violation, incomplete)

Second iteration: Add a "master toggle" in the top right-hand corner of the panel that shows all/hides all highlight borders for that tab. This would also set all of the rule toggles to either on or off. (I think this should default to hiding so that the screen is not cluttered - especially in success it will be)

If we want to add more to this feature in a third iteration, we can do so after the first 2 are released. Ideally, we can get feedback about the feature which would drive design improvements.

What do you think?

**Comment by domyen** [comment]
@CodeByAlex 
- Agree with your assessment of the hover state 👍 
- Agree that highlights should be hidden by default. 

I'm on the fence about individually toggling the highlights for each rule. That level of specificity might be too fine for v1. As a developer I'm concerned by which areas of my UI are in violation. Visualizing them all at once would solve most of my use case.

**How about reversing the order of operations:**
1. Start with the master highlight toggle 
2. Refine with individual toggles for each rule (or some other way to call attention to elements)

Either way, I'm stoked by this initiative!

**Comment by CodeByAlex** [comment]
Hey guys! Attached are some progress snapshots and a screen capture. @domyen, I ended up going from sub-rule level up during development.  Let me know if you think that there are some design modifications that should be made (BTW the highlighting is using 2px instead of 1px dotted outline because I felt that it was easier to spot out).

At this point, Im working to finish the feature in one release instead of two parts. Of course we can look for feedback after that. 

Since the toggles have to know about each other's state and act accordingly the work was a little more complicated than expected. I'm still working through some bugs and I have to get the tab toggle in place. The current a11y panel design and flow of data don't easily allow me to place the highlight toggle component in the tab bar so I'll have to do some code refactoring to make that happen.

Let me know what you think so far!

![image](https://user-images.githubusercontent.com/29763590/54530223-96f37c80-4958-11e9-892d-bc1bf885e9db.png)
![image](https://user-images.githubusercontent.com/29763590/54530255-abd01000-4958-11e9-96bc-0514e3d99099.png)
![highlight-feature-progress](https://user-images.githubusercontent.com/29763590/54530373-f81b5000-4958-11e9-9bea-6e46e2e63f6b.gif)

@shilman @Armanio

**Comment by Armanio** [comment]
Wow! Thank you, @CodeByAlex, for your effort!
Hmm, I think that one toggle is better because many toggles can increase complexity. 🤔 
Anyway, I like your concept!

**Comment by CodeByAlex** [comment]
@Armanio The way I was looking at it was that the rule toggle would be able to show everything for that rule but if you wanted to see what was going wrong with a fine-grained part of that rule, it might be good for the user to have the ability to see where that issue is located.

The complexity is pretty much located in one area.
High-level design:

Highlight-Toggle component - can take in one to many elements that it can highlight. This means that the toggle component can be placed at any level and highlight the components or component passed in based on a switch.

Redux is being used to keep track of a global map where the key is the element itself and the value is an object containing the toggle state of the element, the original outline that the element had so that the component isn't changed when the highlight is removed, and the rule type.

The highlight toggle uses the redux map to keep track of whether or not the element or elements within it are highlighted and whether or not the toggle state is true.

**Comment by domyen** [comment]
@CodeByAlex this is great stuff 🤗

**Feedback**
- The 2px width outlines are pretty in-your-face, especially when there are a lot of highlights on screen. The presence of a red outline/green/yellow outline is quite obvious so I'd still suggest 1px width. 
- In case you weren't already using them, there are theme color variables for consistency. `theme.color.negative`, `theme.color.positive`, `theme.color.warning`

**Comment by CodeByAlex** [comment]
@domyen thanks for the feedback. Ill reduce the px to 1px and make sure to use the theme colors. What do you think about the look and feel of the toggle?

**Comment by domyen** [comment]
@CodeByAlex The switch UI element doesn't give the user much info to predict what will happen if toggled. "What exactly is turning on and off?"

I think it would be more obvious if the toggle appeared as a text link instead of a "switch". For instance, an link that read "Show"/"Hide". Not as handsome, but much more obvious. What do you think?

**Comment by CodeByAlex** [comment]
@domyen, I agree, the switch itself doesn't tell us much about the action. I was able to add the 3rd tier which I labeled but that may not be enough still.
![image](https://user-images.githubusercontent.com/29763590/54622689-5ff89600-4a40-11e9-88d3-c4eb5d5c0b43.png)

What do you think about doing something like this with show/hide rather than on/off text: https://codepen.io/felipefialho/pen/DIdKj

**Comment by domyen** [comment]
**Re: 1px highlight**
Much nicer while remaining super obvious. 👍 

**Re: toggle**
I like what you linked if we were only dealing with on/off, but the iOS toggle feels like we're forcing a UI pattern here. Perhaps a checkbox is simpler while being just as visually recognizable and self-evident. 
<img width="510" alt="Screen Shot 2019-03-19 at 2 01 05 PM" src="https://user-images.githubusercontent.com/263385/54630110-78bc7800-4a4f-11e9-9536-1bc9f3907d1f.png">

**Highlight all**
Love the highlight results toggle! How about changing the UI to something like this:
<img width="973" alt="Screen Shot 2019-03-19 at 1 50 48 PM" src="https://user-images.githubusercontent.com/263385/54629487-08f9bd80-4a4e-11e9-9d36-916ce6acc46d.png">
- Positioned in the subheader to conserve vertical space and is consistent with placing actions to the right of tabs that we already use in SB
- Uses a regular checkbox 


Note: Would love to redesign how the accessibility result items work someday! For now, my suggestions are necessarily within the constraints of the current design. 😅

**Comment by CodeByAlex** [comment]
The PR is out! You can find it here: #6218

**Comment by shilman** [comment]
Shiver me timbers!! I just released https://github.com/storybooks/storybook/releases/tag/v5.1.0-alpha.15 containing PR #6218 that references this issue. Upgrade today to try it out!

Because it's a pre-release you can find it on the `@next` NPM tag.

Closing this issue. Please re-open if you think there's still more to do.
